src/compiler/tsc.ts

```diff
@@ -387,28 +371,50 @@ namespace ts {
-                    sys.exit(ExitStatus.DiagnosticsPresent_OutputsSkipped);
-                }
+            return configParseResult;
+        }
 
-                if (!directoryWatcher && sys.watchDirectory && configFileName) {
-                    const directory = ts.getDirectoryPath(configFileName);
-                    directoryWatcher = sys.watchDirectory(
-                        // When the configFileName is just "tsconfig.json", the watched directory should be
-                        // the current directory; if there is a given "project" parameter, then the configFileName
-                        // is an absolute file name.
-                        directory == "" ? "." : directory,
-                        watchedDirectoryChanged, /*recursive*/ true);
-                };
+        function setWatchers() {
+            // Report error if watch is unavailable in this host
+            if (!sys.watchFile) {
+                reportDiagnostic(createCompilerDiagnostic(Diagnostics.The_current_host_does_not_support_the_0_option, "--watch"), /* host */ undefined);
+                sys.exit(ExitStatus.DiagnosticsPresent_OutputsSkipped);
+            }
+
+            // Watch the compilation root if we're not already
+            if (!directoryWatcher && sys.watchDirectory && configFileName) {
+                // When the configFileName is just "tsconfig.json", the watched directory should be
+                // the current directory; if there is a given "project" parameter, then the configFileName
+                // is an absolute file name.
+                const directory = ts.getDirectoryPath(configFileName);
+                directoryWatched = ts.getNormalizedAbsolutePath(directory === "" ? "." : directory, sys.getCurrentDirectory());
+                directoryWatcher = sys.watchDirectory(directoryWatched, watchDirectoryChanged, /*recursive*/ true);
+            }
+
+            // Watch the config file if we're not already
+            if (!configFileWatcher && configFileName) {
+                configFileWatcher = sys.watchFile(configFileName, configFileChanged);
+            }
+
+            // See if we need to watch any new additional type roots from outside the project folder
+            if (sys.watchDirectory) {
+                const alreadyWatched = directoryWatched || '';
+                for (let typesRoot of getEffectiveTypeRoots(compilerOptions, sys) || []) {
+                    const typeRootPath = ts.getNormalizedAbsolutePath(typesRoot, sys.getCurrentDirectory());
+                    if (!(typeRootPath in typeRootsWatched) &&
+                            typeRootPath.substr(0, alreadyWatched.length) !== alreadyWatched &&
+                            sys.directoryExists(typeRootPath)) {
+
+                        // Don't need to be recursive since we're just looking for new top-level folders
+                        sys.watchDirectory(typeRootPath, watchDirectoryChanged, /*recursive*/ false);

```

Comment:
do you need to save the watcher (i.e. to dispose it if we want to stop watching this folder)?